src/parser/types.rs

```diff
@@ -578,6 +596,59 @@ impl FunctionDef {
         self.sig.name.value == "main"
     }
 
+    pub fn parse_generics(ctx: &mut ParserCtx, tokens: &mut Tokens) -> Result<HashSet<String>> {
+        let mut generics = HashSet::new();
+
+        // <A, B>
+        // ^
+        if matches!(
+            tokens.peek(),
+            Some(Token {
+                kind: TokenKind::Less,
+                ..
+            })
+        ) {
+            tokens.bump(ctx);
+
+            loop {
+                // <A, B>
+                //  ^
+                let name = tokens.bump_generic(
+                    ctx,
+                    ErrorKind::InvalidFunctionSignature("expected generic parameter"),
+                )?;
+
+                generics.insert(name);
```

Comment:
should error if it already exists in the set

Reply:
this should be handled in another PR which removed the <>